### ISSUE TITLE
Flat Storage refactor

### ISSUE THREAD

**Issue by pugachAG** [near/nearcore#8577]
### Priorities

* **P0**: should be implemented before MVP rollout. Involves addressing major pain points as well as changes to persistent data format (to avoid migration overhead in the near future).
* **P1**: important, but not a blocker for MVP. Can be implemented in parallel with MVP rollout.
* **P2**: nice-to-have improvements.

### Proposed changes

* [x] get rid of excessive `#[cfg(feature = "protocol_feature_flat_state")]` usage
  - P0
  - #8668
  - we still need to for `DBCol::FlatState*` to avoid creating rocksdb column when flat storage is not enabled
  - use it to actually enable flat storage at the top level
* [x] split code into multiple smaller files
  - P0
  - #8664
  - add `flat` module under `core/store/src/`, so we have everything related to flat storage in `near_store::flat` module
  - use separate sub-modules under `near_store::flat` for deltas, chunk_view, helpers, etc.
* [x] rename structs
  - P0
  - #8664
  - `FlatStorageState` -> `FlatStorage`
  - `FlatState` -> `FlatStorageChunkView`
  - `FlatStorageFactory` -> `FlatStorageManager` (also considered `GlobalFlatStorage` | `FlatStorageContainer` | `FlatStorageHandler` | `FlatStoragePool` | `ShardFlatStorages`)
* [x] Get rid of `ChainAccessForFlatStorage` as it results in very cumbersome flat storage creation and makes it hard to reason about flat storage consistency from the perspective of the chain.
  - P0: involves data format changes
  - #8683
  - Introduce `FlatStateDeltaMetadata` column and store `BlockInfo` as part of that.  In the future it can be used to store more metadata related to flat state deltas.
  - Also see [this comment](https://github.com/near/nearcore/issues/8577#issuecomment-1435018521).
* [x] Prepend ShardUId to flat storage-related keys on disk
  - P0
  - For FlatState: https://github.com/near/nearcore/issues/8670
  - For deltas: https://github.com/near/nearcore/pull/8710
  - Replace `shard_id` with `shard_uid` for status
  - #8745
* [x] Remove in-memory LRU cache, hit rate is very low
  - P0
  - #8727 
* [x] Unify flat storage states
  - P0: involves data format changes
  - #8689
  - Introduce `FlatStorageState` db column and corresponding enum with the following states:
    - `Disabled` instead of `FlatStorageCreationStatus::DontCreate`
    - `Ready(FlatStorageHead)` to replace `FlatStorageCreationStatus::Ready` and store flat storage head
    - `Creation(FlatStorageCreationStatus)`
    - `Empty` to explicitly express flat storage not having any state for the shard
* [x] other
  - #8752
 
### Implementation

Implementing all improvements listed above one at a time on top of the current code would enormous amount time. So instead I suggest adding new code to `near_store::flat` module without modifying existing implementation. This way we can quickly prototype the new interface (using `todo!()` for implementation) to see if everything fits well together and then we move implementation details from current modules to the new ones. This way only need to spend time on fixing annoying test failures once instead of doing it on every step of the way, which proven to be very time-consuming in the past.
**Update:** unfortunately that didn't work out, falling back to implementing this with small incremental changes.

**Comment by Longarithm** [comment]
I talked to Min a bit, and now I think that we not just need `RuntimeAdapter::notify_block` and `RuntimeAdapter::notify_chunk`, but also `RuntimeAdapter::add_shard` and `RuntimeAdapter::remove_shard`… because how else NightshadeRuntime will know that Chain started/stopped tracking a shard? And we need to pass new shard state to it somehow.

Another impact of flat storage I didn’t think before is that it makes NightshadeRuntime stateful (we start to assume that it keeps info about some blocks/chunks) but the current philosophy is that it is stateless - we can get a trie or return results of applying txns, but it doesn’t impact any state. Don’t know how much we care though.

Minor note - I'm all for renaming "fetching_state" to "writing_state" or smth like that, because writing the state is the final goal.

**Comment by Longarithm** [comment]
Some consensus we reached today - we introduce `FlatStateDeltaMetadata` storing `BlockInfo` so if we create any `FlatStorage` instance, it is not necessary to have existing `Chain` (to create `ChainAccessForFlatStorage`).

We didn't want that before because we found it wrong to maintain duplicated Chain data. However, `FlatStorage` dependency on `Chain` is annoying if we want to create it for tests, debug tools, etc. It may be a bit tricky to implement flat storage creation without such dependency, but at least we can separate this logic and send state parts to flat storage without knowledge how these parts were obtained.

cc @mzhangmzz if you have strong opinion against it :)

**Comment by walnut-the-cat** [comment]
Feedback: This issue is missing:
1. why we want to do these works. (What are the current pain points?)
2. what the high level timeline is. (how long will it take to complete the work? how do we plan to tackle them?)
3. which ones are blockers and which ones are good to have (P0 vs P1/P2)
4.

**Comment by Longarithm** [comment]
I think we can close it, as we solved all P0s needed for flat storage MVP and it is good to go for 1.34 release.
Moving rest of the tasks to #8826, because we currently plan to continue improving flat storage code in Q2.
